simple-generated/src/main.rs

```diff
@@ -1,45 +1,111 @@
-use std::io;
+use std::time::Duration;
 
-use ratatui::{backend::CrosstermBackend, Terminal};
-
-use crate::{
-    app::{App, AppResult},
-    event::{Event, EventHandler},
-    handler::handle_key_events,
-    tui::Tui,
+use color_eyre::Result;
+use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
+use event::{Event, EventSource};
+use ratatui::{
+    layout::Alignment,
+    style::{Style, Stylize},
+    widgets::{Block, BorderType, Paragraph},
+    DefaultTerminal, Frame,
 };
 
-pub mod app;
 pub mod event;
-pub mod handler;
-pub mod tui;
-pub mod ui;
-
-fn main() -> AppResult<()> {
-    // Create an application.
-    let mut app = App::new();
-
-    // Initialize the terminal user interface.
-    let backend = CrosstermBackend::new(io::stdout());
-    let terminal = Terminal::new(backend)?;
-    let events = EventHandler::new(250);
-    let mut tui = Tui::new(terminal, events);
-    tui.init()?;
-
-    // Start the main loop.
-    while app.running {
-        // Render the user interface.
-        tui.draw(&mut app)?;
-        // Handle events.
-        match tui.events.next()? {
-            Event::Tick => app.tick(),
-            Event::Key(key_event) => handle_key_events(key_event, &mut app)?,
-            Event::Mouse(_) => {}
-            Event::Resize(_, _) => {}
+
+fn main() -> Result<()> {
```

Comment:
Why `main` is at the top of the file?

Reply:
;)
bottom
to
top
read
we
Because

(not the other way around)

Reply:
fwiw I dislike the top to bottom pattern. I tend to write code the way you would have to in C without explicit declarations, i.e. any function that needs to be used has to be defined above the function in which it is used. I also use this for structs, imports etc. And when I write it like this `main` always is at the end of the file. 

There are two reasons I dislike the top to bottom pattern, 1) it can the `main` somewhere in the middle of the file. 2) I don’t read code top to bottom sequentially in a file. I usually read it by tracing execution, i.e. I open the bottom of the file, scroll into `main`, use LSP to jump the definition of a function I’m interested in, jump back to main when I’m done, proceed.

Reply:
My reading order is find main and start reading in the same fashion as yours. It doesn't matter to me much where main starts. Cmd+F or Cmd+T works find. But I have an expectation that structs and enums are defined before their implementation. And I significantly dislike having to read functions and methods that are implemented above the code that calls them. When this happens, it feels like having to refer to a footnote in a book that's defined several pages previous. This interrupts the flow of my reading. It's not a linear top to bottom reading thing, but a cognitive burden thing. Functions and methods defined earlier in a file have a higher abstraction, more general, and have shorter names to reflect this (e.g. `main(), run(), init()`), while those lower in a file are more concrete and specific (`on_key_press()`, `increment_counter()`). I understand why C and other languages made this style (I learnt C++ as a kid in the 80s), but there are significant enough benefits of this top down approach that I think are worth always adhering to.

This practice generally is coupled with intention revealing / implementation hiding names for functions, which if done properly makes it easier to see the structure of an app.

I know in the past we've had similar discussions about:
```rust
fn foo() {
    // a
    some code related to a
    some code related to a
    some code related to a

    // b
    some code related to b
    some code related to b
    some code related to b

    // c
    some code related to c
    some code related to c
    some code related to c
}

/// vs
fn foo() {
    a();
    b();
    c();
}

fn a() {...}
fn b() {...}
fn c() {...}
```

This falls in the same bucket for me. Pulling out the code into methods puts a natural boundary around the possible interactions between each block of code. Arranging them in the calling order helps make that feel natural when reading too.
    
I'm definitely an opinionated datapoint of 1 here, but I've found that source code arranged this way seems to be the type that I find is significantly quicker to read diagonally. And it's because it's arranged like a good article is, the most important stuff is up front, the supporting stuff follows after. My metric really is based purely on how long it takes to understand the code when reading it later and how much of the code that I can skip reading rather than reading every single line of every single function.

Reply:
Minor rebuttal - I like `struct`s and `enum`s defined before they are used for the same reason I like functions to be defined before they are used. It's like the building blocks to form a program. Reading the file top to bottom builds cognitively the necessary information to understand what the program can do. 

I do agree that what you are saying has advantages too. At the end of the day it's probably two different styles? 

fwiw, I think writing code in "definition always before use" is simpler to follow. Also, for me another advantage is that it works in every language (python, C, C++, rust, nim, Julia etc) and I don't have to give it any more thought. I personally wouldn't choose a style that orders functions one way in C / C++ / nim and another way in Python, Rust, Julia.

Reply:
All that said, I'm indifferent to what approach we choose here! Most of my comments were related to reading `examples` code where everything is typically in one file, and in this case I think both ways are fine, since we are splitting content across files too.